src/main/java/com/lorepo/icplayer/client/module/choice/ChoicePrintable.java

```diff
@@ -21,49 +22,117 @@ public ChoicePrintable(ChoiceModel model, PrintableController controller) {
 	
 	public String getPrintableHTML(String className, boolean showAnswers) {
 		if (model.getPrintable() == PrintableMode.NO) return null;
-		
-		ArrayList<ChoiceOption> orderedOptions = new ArrayList<ChoiceOption>();
-		for (int i = 0; i < model.getOptionCount(); i++) {
-			orderedOptions.add(model.getOption(i));
+
+		HashMap<String, String> printableState = model.getPrintableState();
+		List<String> optionsHTMLs;
+		if (printableState != null && printableState.containsKey("options")) {
+			optionsHTMLs = getStyledOptionsIfPrintableState(showAnswers, printableState);
 		}
-		
+		else {
+			optionsHTMLs = getStyledOptionsIfNotPrintableState(showAnswers);
+		}
+
 		if (model.isRandomOrder()) {
-			for(int index = 0; index < orderedOptions.size(); index += 1) {  
-			    Collections.swap(orderedOptions, index, index + nextInt(orderedOptions.size() - index));  
+			for(int index = 0; index < optionsHTMLs.size(); index += 1) {
+				Collections.swap(optionsHTMLs, index, index + Random.nextInt(optionsHTMLs.size() - index));
 			}
 		}
-		
-		List<String> optionHTMLs = new ArrayList<String>();
-		for (ChoiceOption option: orderedOptions) {
-			String optionHTML = createPrintableOption(option, showAnswers);
-			optionHTMLs.add(optionHTML);
-		}
-		
+
 		String result = "<div class=\"printable_ic_choice\" id=\"" + model.getId() +"\"><ol type=\"A\">";

```

Comment:
Problem z konkatenacja stringow z palca jest takie, ze ciezko to testowac i ciezko to reuzywac. 
Juz nie mowiac o tym ze np. options tworzysz jako liste jego htmli. 

Defacto w kodzie powinien byc model w stylu OptionsCollection / cos podobnego i na nim tylko wywolac .toXML(), a wewnatrz na podobiektach w liscie .toXML()

Operacje an xmlu powinny byc kodowe - patrz icplayer content parser.

Wtedy duzo latwiej jest napisac testy do takiego kawalka kodu i weryfikowac czy dal danych typow funkcjonalnosci np. odpowiednie funkcje przechodza. Dodatkowo latow jest przygotowac szablon dobrej odpowiedzi (plik z xml'em wczytywany do testu).

Reply:
Został zmieniony kod z palca na elementy DOM. 
Ze względu na różne rodzaje klas elementów (XML, DOM) oraz na większe możliwości zrozumienia kodu, to nie została dodana funkcja na wzór toXML bezpośrednio w klasie option. Rozchodzi się o to, że printowalnej wersji nie korzysta się z wszystkich danych z obiektu option. Konkretniej to korzysta się ze stanów, które są pobierane przy setPrintableState/getPrintableState a nie ze stanu w obiektach. Dlatego dodałem dodatkową klasę, która odpowiada za printowalne wersji dla obiektu opcji.